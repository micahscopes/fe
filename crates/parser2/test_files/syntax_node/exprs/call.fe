foo()
foo::Bar()
foo(x: 1, z: 3)
foo(x: 1, z: 3)
foo(x: 1, 2, z: 3)
foo(1, y: 2, z: 3)

foo<i32, foo::MyType>(val1: 2, val2: "String")
foo<[u32; 1], {3 + 4}>(x: 1, y: 2)

foo::bar<T, U>(x)

// Ths should be parsed as `(foo<i32, (u32)>(1))`, not a tuple expression.
(foo < i32, (u32) > (1))
