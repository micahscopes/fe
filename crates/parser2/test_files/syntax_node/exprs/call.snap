---
source: crates/parser2/tests/syntax_node.rs
expression: node
input_file: crates/parser2/test_files/syntax_node/exprs/call.fe
---
Root@0..290
  CallExpr@0..5
    PathExpr@0..3
      Path@0..3
        PathSegment@0..3
          Ident@0..3 "foo"
    CallArgList@3..5
      LParen@3..4 "("
      RParen@4..5 ")"
  Newline@5..6 "\n"
  CallExpr@6..16
    PathExpr@6..14
      Path@6..14
        PathSegment@6..9
          Ident@6..9 "foo"
        Colon2@9..11 "::"
        PathSegment@11..14
          Ident@11..14 "Bar"
    CallArgList@14..16
      LParen@14..15 "("
      RParen@15..16 ")"
  Newline@16..17 "\n"
  CallExpr@17..32
    PathExpr@17..20
      Path@17..20
        PathSegment@17..20
          Ident@17..20 "foo"
    CallArgList@20..32
      LParen@20..21 "("
      CallArg@21..25
        Ident@21..22 "x"
        Colon@22..23 ":"
        WhiteSpace@23..24 " "
        LitExpr@24..25
          Lit@24..25
            Int@24..25 "1"
      Comma@25..26 ","
      WhiteSpace@26..27 " "
      CallArg@27..31
        Ident@27..28 "z"
        Colon@28..29 ":"
        WhiteSpace@29..30 " "
        LitExpr@30..31
          Lit@30..31
            Int@30..31 "3"
      RParen@31..32 ")"
  Newline@32..33 "\n"
  CallExpr@33..48
    PathExpr@33..36
      Path@33..36
        PathSegment@33..36
          Ident@33..36 "foo"
    CallArgList@36..48
      LParen@36..37 "("
      CallArg@37..41
        Ident@37..38 "x"
        Colon@38..39 ":"
        WhiteSpace@39..40 " "
        LitExpr@40..41
          Lit@40..41
            Int@40..41 "1"
      Comma@41..42 ","
      WhiteSpace@42..43 " "
      CallArg@43..47
        Ident@43..44 "z"
        Colon@44..45 ":"
        WhiteSpace@45..46 " "
        LitExpr@46..47
          Lit@46..47
            Int@46..47 "3"
      RParen@47..48 ")"
  Newline@48..49 "\n"
  CallExpr@49..67
    PathExpr@49..52
      Path@49..52
        PathSegment@49..52
          Ident@49..52 "foo"
    CallArgList@52..67
      LParen@52..53 "("
      CallArg@53..57
        Ident@53..54 "x"
        Colon@54..55 ":"
        WhiteSpace@55..56 " "
        LitExpr@56..57
          Lit@56..57
            Int@56..57 "1"
      Comma@57..58 ","
      WhiteSpace@58..59 " "
      CallArg@59..60
        LitExpr@59..60
          Lit@59..60
            Int@59..60 "2"
      Comma@60..61 ","
      WhiteSpace@61..62 " "
      CallArg@62..66
        Ident@62..63 "z"
        Colon@63..64 ":"
        WhiteSpace@64..65 " "
        LitExpr@65..66
          Lit@65..66
            Int@65..66 "3"
      RParen@66..67 ")"
  Newline@67..68 "\n"
  CallExpr@68..86
    PathExpr@68..71
      Path@68..71
        PathSegment@68..71
          Ident@68..71 "foo"
    CallArgList@71..86
      LParen@71..72 "("
      CallArg@72..73
        LitExpr@72..73
          Lit@72..73
            Int@72..73 "1"
      Comma@73..74 ","
      WhiteSpace@74..75 " "
      CallArg@75..79
        Ident@75..76 "y"
        Colon@76..77 ":"
        WhiteSpace@77..78 " "
        LitExpr@78..79
          Lit@78..79
            Int@78..79 "2"
      Comma@79..80 ","
      WhiteSpace@80..81 " "
      CallArg@81..85
        Ident@81..82 "z"
        Colon@82..83 ":"
        WhiteSpace@83..84 " "
        LitExpr@84..85
          Lit@84..85
            Int@84..85 "3"
      RParen@85..86 ")"
  Newline@86..88 "\n\n"
  CallExpr@88..134
    PathExpr@88..109
      Path@88..109
        PathSegment@88..109
          Ident@88..91 "foo"
          GenericArgList@91..109
            Lt@91..92 "<"
            TypeGenericArg@92..95
              PathType@92..95
                Path@92..95
                  PathSegment@92..95
                    Ident@92..95 "i32"
            Comma@95..96 ","
            WhiteSpace@96..97 " "
            TypeGenericArg@97..108
              PathType@97..108
                Path@97..108
                  PathSegment@97..100
                    Ident@97..100 "foo"
                  Colon2@100..102 "::"
                  PathSegment@102..108
                    Ident@102..108 "MyType"
            Gt@108..109 ">"
    CallArgList@109..134
      LParen@109..110 "("
      CallArg@110..117
        Ident@110..114 "val1"
        Colon@114..115 ":"
        WhiteSpace@115..116 " "
        LitExpr@116..117
          Lit@116..117
            Int@116..117 "2"
      Comma@117..118 ","
      WhiteSpace@118..119 " "
      CallArg@119..133
        Ident@119..123 "val2"
        Colon@123..124 ":"
        WhiteSpace@124..125 " "
        LitExpr@125..133
          Lit@125..133
            String@125..133 "\"String\""
      RParen@133..134 ")"
  Newline@134..135 "\n"
  CallExpr@135..169
    PathExpr@135..157
      Path@135..157
        PathSegment@135..157
          Ident@135..138 "foo"
          GenericArgList@138..157
            Lt@138..139 "<"
            TypeGenericArg@139..147
              ArrayType@139..147
                LBracket@139..140 "["
                PathType@140..143
                  Path@140..143
                    PathSegment@140..143
                      Ident@140..143 "u32"
                SemiColon@143..144 ";"
                WhiteSpace@144..145 " "
                LitExpr@145..146
                  Lit@145..146
                    Int@145..146 "1"
                RBracket@146..147 "]"
            Comma@147..148 ","
            WhiteSpace@148..149 " "
            ConstGenericArg@149..156
              BlockExpr@149..156
                LBrace@149..150 "{"
                ExprStmt@150..155
                  BinExpr@150..155
                    LitExpr@150..151
                      Lit@150..151
                        Int@150..151 "3"
                    WhiteSpace@151..152 " "
                    Plus@152..153 "+"
                    WhiteSpace@153..154 " "
                    LitExpr@154..155
                      Lit@154..155
                        Int@154..155 "4"
                RBrace@155..156 "}"
            Gt@156..157 ">"
    CallArgList@157..169
      LParen@157..158 "("
      CallArg@158..162
        Ident@158..159 "x"
        Colon@159..160 ":"
        WhiteSpace@160..161 " "
        LitExpr@161..162
          Lit@161..162
            Int@161..162 "1"
      Comma@162..163 ","
      WhiteSpace@163..164 " "
      CallArg@164..168
        Ident@164..165 "y"
        Colon@165..166 ":"
        WhiteSpace@166..167 " "
        LitExpr@167..168
          Lit@167..168
            Int@167..168 "2"
      RParen@168..169 ")"
  Newline@169..171 "\n\n"
  CallExpr@171..188
    PathExpr@171..185
      Path@171..185
        PathSegment@171..174
          Ident@171..174 "foo"
        Colon2@174..176 "::"
        PathSegment@176..185
          Ident@176..179 "bar"
          GenericArgList@179..185
            Lt@179..180 "<"
            TypeGenericArg@180..181
              PathType@180..181
                Path@180..181
                  PathSegment@180..181
                    Ident@180..181 "T"
            Comma@181..182 ","
            WhiteSpace@182..183 " "
            TypeGenericArg@183..184
              PathType@183..184
                Path@183..184
                  PathSegment@183..184
                    Ident@183..184 "U"
            Gt@184..185 ">"
    CallArgList@185..188
      LParen@185..186 "("
      CallArg@186..187
        PathExpr@186..187
          Path@186..187
            PathSegment@186..187
              Ident@186..187 "x"
      RParen@187..188 ")"
  Newline@188..190 "\n\n"
  Comment@190..264 "// Ths should be pars ..."
  Newline@264..265 "\n"
  ParenExpr@265..289
    LParen@265..266 "("
    CallExpr@266..288
      PathExpr@266..284
        Path@266..284
          PathSegment@266..284
            Ident@266..269 "foo"
            WhiteSpace@269..270 " "
            GenericArgList@270..284
              Lt@270..271 "<"
              WhiteSpace@271..272 " "
              TypeGenericArg@272..275
                PathType@272..275
                  Path@272..275
                    PathSegment@272..275
                      Ident@272..275 "i32"
              Comma@275..276 ","
              WhiteSpace@276..277 " "
              TypeGenericArg@277..282
                TupleType@277..282
                  LParen@277..278 "("
                  PathType@278..281
                    Path@278..281
                      PathSegment@278..281
                        Ident@278..281 "u32"
                  RParen@281..282 ")"
              WhiteSpace@282..283 " "
              Gt@283..284 ">"
      WhiteSpace@284..285 " "
      CallArgList@285..288
        LParen@285..286 "("
        CallArg@286..287
          LitExpr@286..287
            Lit@286..287
              Int@286..287 "1"
        RParen@287..288 ")"
    RParen@288..289 ")"
  Newline@289..290 "\n"

